### PR TITLE
DB-8793 Remove deleted rows during compaction

### PR DESCRIPTION
**Pull request by arnaud-splice**
We can remove deleted rows only if active transactions are using a more
recent version of that row.

In the event of a minor compaction, since we do not see all hfiles and
might have a partial view of the history of a given row, we delete
everything BUT the tombstone itself. This way, the portion of history
that lived in a different hfile will still fall earlier than the
remaining tombstone, ensuring that the known state of the row is
"deleted".
Those remaining tombstones shall be removed during major compaction.

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please retest branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

File: hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionState.java [hunk 1]
```diff
@@ -69,28 +70,30 @@ public SICompactionState(TxnSupplier transactionStore, int activeTransactionCach
...
+    public void mutate(List<Cell> rawList, List<TxnView> txns, List<Cell> results, boolean forcePurgingDeletedRows, boolean keepTombstones) throws IOException {
         dataToReturn.clear();
+        long lowWatermarkTransaction = TransactionsWatcher.INSTANCE.getLowWatermarkTransaction();
         long maxTombstone = 0;
         Iterator<TxnView> it = txns.iterator();
         for (Cell aRawList : rawList) {
             TxnView txn = it.next();
             long t = mutate(aRawList, txn);
-            if (t > maxTombstone) {
+            if (maxTombstone < t && t < lowWatermarkTransaction)
```
Comment (dgomezferro) [inline hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionState.java]: Here I think we are processing all tombstones, not just only the committed ones. Tombstones from Active transactions should be ignored due to the lowWatermarkTransaction check, but I think we'd be removing data based on rolledback tombstones.

Comment (arnaud-splice) [reply hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionState.java]: Good catch. That should be addressed in 807e48d

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Pull request closed without merging**